src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -396,7 +405,8 @@ void GcsTaskManager::HandleGetTaskEvents(rpc::GetTaskEventsRequest request,
   int64_t num_limit_truncated = 0;
 
   // A lambda filter fn, where it returns true for task events to be included in the
-  // result. Task ids and job ids are already filtered by the storage with indexing above.
+  // result. Task ids or job ids with equal predicate are already filtered by the
+  // storage with indexing above.
```

Comment:
I tried to reduce some boilerplate by doing this:

1. define `apply_op` and alike to treat the ==EQUAL or ==NOT_EQUAL thing
2. use it in filter_fn.

Please see if this makes sense. (did not build it, you may want to tweak a bit)



```cpp

template <typename T>
bool apply_op(const T& lhs, rpc::FilterPredicate op, const T& rhs) {
  switch (op) {
  case rpc::FilterPredicate::EQUAL:
    return lhs == rhs;
  case rpc::FilterPredicate::NOT_EQUAL:
    return lhs != rhs;
  default:
    RAY_LOG(FATAL) << "Unknown filter predicate: " << op;
  }
}


bool apply_op_ignore_case(std::string_view lhs,
                          rpc::FilterPredicate op,
                          std::string_view rhs) {
  bool equals = absl::EqualsIgnoreCase(lhs, rhs);
  switch (op) {
  case rpc::FilterPredicate::EQUAL:
    return equals;
  case rpc::FilterPredicate::NOT_EQUAL:
    return !equals;
  default:
    RAY_LOG(FATAL) << "Unknown filter predicate: " << op;
  }
}
```

```cpp

  // A lambda filter fn, where it returns true for task events to be included in the
  // result. Task ids and job ids are already filtered by the storage with indexing above.
  auto filter_fn = [&filters](const rpc::TaskEvents &task_event) {
    if (!task_event.has_task_info()) {
      // Skip task events w/o task info.
      return false;
    }
    if (filters.exclude_driver() &&
        task_event.task_info().type() == rpc::TaskType::DRIVER_TASK) {
      return false;
    }

    // Job filter
    if (filters.has_job_filter() && !apply_op(
        JobID::FromBinary(task_event.job_id()),
        filters.job_filter().predicate(),
        JobID::FromBinary(filters.job_filter().job_id()))) {
      return false;
    }

    // Tasks filters (OR semantics)
    if (filters.task_filters_size() > 0) {
      if (!std::any_of(filters.task_filters().begin(),
                       filters.task_filters().end(),
                       [&task_event](const auto &task_filter) {
                         return apply_op(TaskID::FromBinary(task_event.task_id()),
                                       task_filter.predicate(),
                                       TaskID::FromBinary(task_filter.task_id()));
                       })) {
        return false;
      }
    }

    // Actor filter
    if (filters.has_actor_filter() && task_event.task_info().has_actor_id() &&
        !apply_op(ActorID::FromBinary(task_event.task_info().actor_id()),
                  filters.actor_filter().predicate(),
                  ActorID::FromBinary(filters.actor_filter().actor_id()))) {
      return false;
    }

    // Name filter
    if (filters.has_name_filter()) {
      if (!apply_op_ignore_case(
              task_event.task_info().name(),
              filters.name_filter().predicate(),
              filters.name_filter().name())) {
        return false;
      }
    }

    // State filter
    if (filters.has_state_filter()) {
      const google::protobuf::EnumDescriptor *task_status_descriptor =
          ray::rpc::TaskStatus_descriptor();

      // Figure out the latest state of a task
      ray::rpc::TaskStatus state = ray::rpc::TaskStatus::NIL;
      if (task_event.has_state_updates()) {
        for (int i = task_status_descriptor->value_count() - 1; i >= 0; --i) {
          if (task_event.state_updates().state_ts_ns().contains(
                  task_status_descriptor->value(i)->number())) {
            state = static_cast<ray::rpc::TaskStatus>(
                task_status_descriptor->value(i)->number());
            break;
          }
        }
      }

      if (!apply_op_ignore_case(
              task_status_descriptor->FindValueByNumber(state)->name(),
              filters.state_filter().predicate(),
              filters.state_filter().state())) {
        return false;
      }
    }

    return true;
  };

```